spec/spec_helper.cr

```diff
@@ -97,28 +75,38 @@ def test_headers(headers = nil)
   req_hdrs
 end
 
-def start_amqp_server
+def with_amqp_server(tls = false, & : LavinMQ::Server -> Nil)
+  tcp_server = TCPServer.new("::1", 0)
   s = LavinMQ::Server.new(DATA_DIR)
-  tcp_server = TCPServer.new(LavinMQ::Config.instance.amqp_bind, 0)
-  LavinMQ::Config.instance.amqp_port = tcp_server.local_address.port
-  spawn { s.listen(tcp_server) }
-  spawn { s.listen_clustering("127.0.0.1", LavinMQ::Config.instance.clustering_port) }
-  ctx = OpenSSL::SSL::Context::Server.new
-  ctx.certificate_chain = "spec/resources/server_certificate.pem"
-  ctx.private_key = "spec/resources/server_key.pem"
-  tls_server = TCPServer.new(LavinMQ::Config.instance.amqp_bind, 0)
-  LavinMQ::Config.instance.amqps_port = tls_server.local_address.port
-  spawn { s.listen_tls(tls_server, ctx) }
-  s
+  begin
+    if tls
+      ctx = OpenSSL::SSL::Context::Server.new
+      ctx.certificate_chain = "spec/resources/server_certificate.pem"
+      ctx.private_key = "spec/resources/server_key.pem"
+      spawn(name: "amqp tls listen") { s.listen_tls(tcp_server, ctx) }
+    else
+      spawn(name: "amqp tcp listen") { s.listen(tcp_server) }
+    end
+    Fiber.yield
+    yield s
+  ensure
+    s.close
+    FileUtils.rm_rf("/tmp/lavinmq-spec")
+  end
 end
 
-def start_http_server
-  h = LavinMQ::HTTP::Server.new(Server)
-  addr = h.bind_tcp(LavinMQ::Config.instance.http_bind, 0)
-  LavinMQ::Config.instance.http_port = addr.port
-  spawn { h.listen }
-  Fiber.yield
-  h
+def with_http_server(&)
+  with_amqp_server do |s|
+    h = LavinMQ::HTTP::Server.new(s)
+    begin
+      addr = h.bind_tcp("::1", 0)
+      spawn(name: "http listen") { h.listen }
+      Fiber.yield
+      yield({addr, s})
```

Comment:
yes, was thinking of that too, let's try that. 

maybe the `with_amqp_server` also should yield a struct/class with a `with_channel` method.

Reply:
a strong benefit of yielding a custom object is that it can be extended without having to rewrite all specs (as opposed as adding a new object that is being yielded).